modules/agar/src/main/ts/ephox/agar/alien/SelectorEngine.ts

```diff
@@ -0,0 +1,59 @@
+import { Arr, Obj, Optional, Strings } from '@ephox/katamari';
+
+type LookupContext = Element | Document | DocumentFragment;
+interface DecodedContainsSelector {
+  baseSelector: string;
+  text: string;
+}
+
+const selectAll = (selector: string, context: LookupContext): Element[] =>
+  decodeContains(selector).fold(
+    () => queryAll(context, selector),
+    (decodedContainsSelector) => queryAllWithContains(context, decodedContainsSelector)
+  );
+
+const decodeContains = (selector: string): Optional<DecodedContainsSelector> => {
+  const regexp = /(?<baseSelector>.*):contains\((?<text>.*)\)$/;
```

Comment:
The problem I have with this is it will fall back to `querySelectorAll` when `:contains` is in the middle - which will fail with `not a valid selector`.

I'd prefer to see this code capture any extra text, and then if that final group isn't empty throw an error that explains what is wrong